src/connector/codec/src/decoder/avro/mod.rs

```diff
@@ -107,6 +108,51 @@ impl<'a> AvroParseOptions<'a> {
 
         let v: ScalarImpl = match (type_expected, value) {
             (_, Value::Null) => return Ok(DatumCow::NULL),
+            // ---- Union (with >=2 non null variants) -----
+            (DataType::Struct(struct_type_info), Value::Union(variant, v)) => match self.schema {
+                Some(Schema::Union(u)) => {
+                    if let Some(inner) = get_nullable_union_inner(u) {
+                        // nullable Union ([null, record])
+                        return Self {
+                            schema: Some(inner),
+                            relax_numeric: self.relax_numeric,
+                        }
+                        .convert_to_datum(v, type_expected);
+                    }
+                    let variant_schema = &u.variants()[*variant as usize];
+
+                    if matches!(variant_schema, &Schema::Null) {
+                        return Ok(DatumCow::NULL);
+                    }
+
+                    // XXX: can we use the variant idx to find the field idx?
+                    // We will need to get the index of the "null" variant, and then re-map the variant index to the field index.
+                    // Which way is better?
```

Comment:
Get you point but the index approach sounds more error prune. Agree to keep current name approach.